In [1]:
import pandas as pd
import numpy as np

In [2]:
df = pd.read_csv("master_data.csv")
df

order_id                       customer_id  \
0       e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   
1       53cdb2fc8bc7dce0b6741e2150273451  b0830fb4747a6c6d20dea0b8c802d7ef   
2       47770eb9100c2d0c44946d9cf07ec65d  41ce2a54c0b03bf3443c3d931a367089   
3       949d5b44dbf5de918fe9c16f97b45f8a  f88197465ea7920adcdbec7375364d82   
4       ad21c59c0840e6cb83a9ceb5573f8159  8ab97904e6daea8866dbdbc4fb7aad2c   
...                                  ...                               ...   
110007  63943bddc261676b46f01ca7ac2f7bd8  1fca14ff2861355f6e5f14306ff977a7   
110008  83c1379a015df1e13d02aae0204711ab  1aa71eb042121263aafbe80c1b562c9c   
110009  11c177c8e97725db2631073c19f07b62  b331b74b18dc79bcdf6532d51e1637c1   
110010  11c177c8e97725db2631073c19f07b62  b331b74b18dc79bcdf6532d51e1637c1   
110011  66dea50a8b16d9b4dee7af250b4be1a5  edb027a75a1449115f6b43211ae02a24   

       order_status order_purchase_timestamp    order_approved_at  \
0         delivered      2017-10-02 10:56:33  2017-10-02 11:07:15   
1         delivered      2018-07-24 20:41:37  2018-07-26 03:24:27   
2         delivered      2018-08-08 08:38:49  2018-08-08 08:55:23   
3         delivered      2017-11-18 19:28:06  2017-11-18 19:45:59   
4         delivered      2018-02-13 21:18:39  2018-02-13 22:20:29   
...             ...                      ...                  ...   
110007    delivered      2018-02-06 12:58:58  2018-02-06 13:10:37   
110008    delivered      2017-08-27 14:46:43  2017-08-27 15:04:16   
110009    delivered      2018-01-08 21:28:27  2018-01-08 21:36:21   
110010    delivered      2018-01-08 21:28:27  2018-01-08 21:36:21   
110011    delivered      2018-03-08 20:57:30  2018-03-09 11:20:28   

       order_delivered_carrier_date order_delivered_customer_date  \
0               2017-10-04 19:55:00           2017-10-10 21:25:13   
1               2018-07-26 14:31:00           2018-08-07 15:27:45   
2               2018-08-08 13:50:00           2018-08-17 18:06:29   
3               2017-11-22 13:39:59           2017-12-02 00:28:42   
4               2018-02-14 19:46:34           2018-02-16 18:17:02   
...                             ...                           ...   
110007          2018-02-07 23:22:42           2018-02-28 17:37:56   
110008          2017-08-28 20:52:26           2017-09-21 11:24:17   
110009          2018-01-12 15:35:03           2018-01-25 23:32:54   
110010          2018-01-12 15:35:03           2018-01-25 23:32:54   
110011          2018-03-09 22:11:59           2018-03-16 13:08:30   

       order_estimated_delivery_date  order_item_id  \
0                         2017-10-18            1.0   
1                         2018-08-13            1.0   
2                         2018-09-04            1.0   
3                         2017-12-15            1.0   
4                         2018-02-26            1.0   
...                              ...            ...   
110007                    2018-03-02            1.0   
110008                    2017-09-27            1.0   
110009                    2018-02-15            1.0   
110010                    2018-02-15            2.0   
110011                    2018-04-03            1.0   

                              product_id  ... customer_state  \
0       87285b34884572647811a353c7ac498a  ...             SP   
1       595fac2a385ac33a80bd5114aec74eb8  ...             BA   
2       aa4383b373c6aca5d8797843e5594415  ...             GO   
3       d0b61bfb1de832b15ba9d266ca96e5b0  ...             RN   
4       65266b2da20d04dbe00c5c2d3bb7859e  ...             SP   
...                                  ...  ...            ...   
110007  f1d4ce8c6dd66c47bbaa8c6781c2a923  ...             SP   
110008  b80910977a37536adeddd63663f916ad  ...             BA   
110009  d1c427060a0f73f6b889a5c7c61f2ac4  ...             RJ   
110010  d1c427060a0f73f6b889a5c7c61f2ac4  ...             RJ   
110011  006619bbed68b000c8ba3f8725d5409e  ...             PR   

       

In [3]:
df.head()

order_id                       customer_id  \
0  e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   
1  53cdb2fc8bc7dce0b6741e2150273451  b0830fb4747a6c6d20dea0b8c802d7ef   
2  47770eb9100c2d0c44946d9cf07ec65d  41ce2a54c0b03bf3443c3d931a367089   
3  949d5b44dbf5de918fe9c16f97b45f8a  f88197465ea7920adcdbec7375364d82   
4  ad21c59c0840e6cb83a9ceb5573f8159  8ab97904e6daea8866dbdbc4fb7aad2c   

  order_status order_purchase_timestamp    order_approved_at  \
0    delivered      2017-10-02 10:56:33  2017-10-02 11:07:15   
1    delivered      2018-07-24 20:41:37  2018-07-26 03:24:27   
2    delivered      2018-08-08 08:38:49  2018-08-08 08:55:23   
3    delivered      2017-11-18 19:28:06  2017-11-18 19:45:59   
4    delivered      2018-02-13 21:18:39  2018-02-13 22:20:29   

  order_delivered_carrier_date order_delivered_customer_date  \
0          2017-10-04 19:55:00           2017-10-10 21:25:13   
1          2018-07-26 14:31:00           2018-08-07 15:27:45   
2          2018-08-08 13:50:00           2018-08-17 18:06:29   
3          2017-11-22 13:39:59           2017-12-02 00:28:42   
4          2018-02-14 19:46:34           2018-02-16 18:17:02   

  order_estimated_delivery_date  order_item_id  \
0                    2017-10-18            1.0   
1                    2018-08-13            1.0   
2                    2018-09-04            1.0   
3                    2017-12-15            1.0   
4                    2018-02-26            1.0   

                         product_id  ... customer_state  \
0  87285b34884572647811a353c7ac498a  ...             SP   
1  595fac2a385ac33a80bd5114aec74eb8  ...             BA   
2  aa4383b373c6aca5d8797843e5594415  ...             GO   
3  d0b61bfb1de832b15ba9d266ca96e5b0  ...             RN   
4  65266b2da20d04dbe00c5c2d3bb7859e  ...             SP   

                          review_id  review_score  review_comment_title  \
0  a54f0611adc9ed256b57ede6b6eb5114           4.0                   NaN   
1  8d5266042046a06655c8db133d120ba5           4.0      Muito boa a loja   
2  e73b67b67587f7644d5bd1a52deb1b01           5.0                   NaN   
3  359d03e676b3c069f62cadba8dd3f6e8           5.0                   NaN   
4  e50934924e227544ba8246aeb3770dd4           5.0                   NaN   

                              review_comment_message  review_creation_date  \
0  Não testei o produto ainda, mas ele veio corre...   2017-10-11 00:00:00   
1                               Muito bom o produto.   2018-08-08 00:00:00   
2                                                NaN   2018-08-18 00:00:00   
3  O produto foi exatamente o que eu esperava e e...   2017-12-03 00:00:00   
4                                                NaN   2018-02-17 00:00:00   

   review_answer_timestamp  delivery_days  is_late  review_binary  
0      2017-10-12 03:43:48              8        0              1  
1      2018-08-08 18:37:50             13        0              1  
2      2018-08-22 19:07:58              9        0              1  
3      2017-12-05 19:21:58             13        0              1  
4      2018-02-18 13:02:51              2        0              1  

[5 rows x 35 columns]

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 110012 entries, 0 to 110011
Data columns (total 35 columns):
 #   Column                         Non-Null Count   Dtype  
---  ------                         --------------   -----  
 0   order_id                       110012 non-null  object 
 1   customer_id                    110012 non-null  object 
 2   order_status                   110012 non-null  object 
 3   order_purchase_timestamp       110012 non-null  object 
 4   order_approved_at              109997 non-null  object 
 5   order_delivered_carrier_date   110011 non-null  object 
 6   order_delivered_customer_date  110012 non-null  object 
 7   order_estimated_delivery_date  110012 non-null  object 
 8   order_item_id                  110012 non-null  float64
 9   product_id                     110012 non-null  object 
 10  seller_id                      110012 non-null  object 
 11  shipping_limit_date            110012 non-null  object 
 12  price                         

In [5]:
df = df[df['order_status'] == 'delivered']

In [6]:
cols = [
    'delivery_days', 'is_late',
    'product_category_name', 'price', 'product_weight_g',
    'freight_value',
    'customer_state',
    'review_binary'
]
df = df[cols]

In [7]:
df['product_category_name'].fillna('Unknown', inplace=True)
df['product_weight_g'].fillna(df['product_weight_g'].median(), inplace=True)

C:\Users\biena\AppData\Local\Temp\ipykernel_16544\437226693.py:1: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df['product_category_name'].fillna('Unknown', inplace=True)
C:\Users\biena\AppData\Local\Temp\ipykernel_16544\437226693.py:2: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

In [8]:
df['delivery_speed'] = pd.cut(
    df['delivery_days'],
    bins=[0,3,7,100],
    labels=['Fast', 'Normal', 'Slow']
)

In [9]:
df['price_level'] = pd.qcut(
    df['price'],
    q=3,
    labels=['Cheap', 'Medium', 'Expensive']
)

In [10]:
df['weight_level'] = pd.qcut(
    df['product_weight_g'],
    q=3,
    labels=['Light', 'Medium', 'Heavy']
)

In [11]:
df['freight_level'] = pd.qcut(
    df['freight_value'],
    q=3,
    labels=['Low', 'Medium', 'High']
)

In [12]:
df['late_item'] = df['is_late'].map({1: 'Late_delivery', 0: 'On_time_delivery'})
df['review_item'] = df['review_binary'].map({1: 'Positive_review', 0: 'Negative_review'})
df['state_item'] = 'State_' + df['customer_state']
df['category_item'] = 'Category_' + df['product_category_name'].astype(str)

In [13]:
transactions = df[[
    'late_item',
    'delivery_speed',
    'price_level',
    'weight_level',
    'freight_level',
    'state_item',
    'category_item',
    'review_item'
]].values.tolist()

In [14]:
transactions

[['On_time_delivery',
  'Slow',
  'Cheap',
  'Medium',
  'Low',
  'State_SP',
  'Category_housewares',
  'Positive_review'],
 ['On_time_delivery',
  'Slow',
  'Expensive',
  'Light',
  'High',
  'State_BA',
  'Category_perfumery',
  'Positive_review'],
 ['On_time_delivery',
  'Slow',
  'Expensive',
  'Medium',
  'High',
  'State_GO',
  'Category_auto',
  'Positive_review'],
 ['On_time_delivery',
  'Slow',
  'Cheap',
  'Medium',
  'High',
  'State_RN',
  'Category_pet_shop',
  'Positive_review'],
 ['On_time_delivery',
  'Fast',
  'Cheap',
  'Light',
  'Low',
  'State_SP',
  'Category_stationery',
  'Positive_review'],
 ['On_time_delivery',
  'Slow',
  'Expensive',
  'Heavy',
  'High',
  'State_PR',
  'Category_auto',
  'Positive_review'],
 ['On_time_delivery',
  'Slow',
  'Medium',
  'Light',
  'Medium',
  'State_RJ',
  'Category_auto',
  'Positive_review'],
 ['On_time_delivery',
  'Slow',
  'Cheap',
  'Light',
  'Medium',
  'State_RS',
  'Category_furniture_decor',
  'Negative_review']

# 1.Những tổ hợp yếu tố liên quan đến giao hàng, đặc điểm sản phẩm, chi phí vận chuyển và khu vực khách hàng có mối liên hệ như thế nào với mức độ hài lòng của khách hàng trên nền tảng thương mại điện tử Olist?

## 1.1. Hiệu suất giao hàng (thời gian giao, giao trễ hay đúng hạn) có liên hệ như thế nào với mức độ hài lòng của khách hàng?

In [15]:
df_g1 = df[[
    'late_item',
    'delivery_speed',
    'review_item'
]].copy()

In [16]:
df['delivery_speed'] = df['delivery_speed'].astype(str)
df['delivery_speed'] = 'Speed_' + df['delivery_speed']

In [17]:
late_review_table = pd.crosstab(
    df_g1['late_item'],
    df_g1['review_item'],
    normalize='index'
) * 100

late_review_table

review_item       Negative_review  Positive_review
late_item                                         
Late_delivery           65.747664        34.252336
On_time_delivery        19.604712        80.395288

* Đơn hàng giao trễ:

Có tới 65.75% bị đánh giá tiêu cực

Chỉ 34.25% đánh giá tích cực

* Đơn hàng giao đúng hạn:

80.40% nhận đánh giá tích cực

Chỉ 19.60% bị đánh giá tiêu cực

=> Hiệu suất giao hàng có mối liên hệ mạnh mẽ với mức độ hài lòng của khách hàng.
Các đơn hàng giao trễ có xác suất nhận đánh giá tiêu cực cao hơn đáng kể so với các đơn giao đúng hạn, trong khi giao đúng hạn có xu hướng gắn liền với mức độ hài lòng cao.

=> giao hàng là một yếu tố cốt lõi ảnh hưởng đến sự hài lòng của khách hàng, tạo nền tảng cho việc tiếp tục phân tích các yếu tố khác như đặc điểm sản phẩm, chi phí vận chuyển và khu vực khách hàng

In [18]:
speed_review_table = pd.crosstab(
    df_g1['delivery_speed'],
    df_g1['review_item'],
    normalize='index'
) * 100

speed_review_table

review_item     Negative_review  Positive_review
delivery_speed                                  
Fast                  14.618409        85.381591
Normal                16.858424        83.141576
Slow                  26.918326        73.081674

* Giao nhanh (Fast):

Tỷ lệ đánh giá tích cực cao nhất (85.38%)

Đánh giá tiêu cực thấp nhất

* Giao bình thường (Normal):

Mức độ hài lòng giảm nhẹ so với giao nhanh

* Giao chậm (Slow):

Tỷ lệ đánh giá tiêu cực tăng mạnh (~27%)

Đánh giá tích cực giảm xuống chỉ còn 73%

=> Có xu hướng rõ ràng:

Tốc độ giao hàng càng chậm → mức độ hài lòng càng giảm

=> Hiệu suất giao hàng, bao gồm cả tình trạng trễ hạn và tốc độ giao hàng, có mối liên hệ mạnh mẽ với mức độ hài lòng của khách hàng trên Olist. Các đơn hàng giao nhanh và đúng hạn có tỷ lệ đánh giá tích cực cao hơn đáng kể so với các đơn giao chậm hoặc giao trễ.

In [19]:
transactions_g1 = df[[
    'late_item',
    'delivery_speed',
    'review_item'
]].values.tolist()

In [20]:
transactions_g1

[['On_time_delivery', 'Speed_Slow', 'Positive_review'],
 ['On_time_delivery', 'Speed_Slow', 'Positive_review'],
 ['On_time_delivery', 'Speed_Slow', 'Positive_review'],
 ['On_time_delivery', 'Speed_Slow', 'Positive_review'],
 ['On_time_delivery', 'Speed_Fast', 'Positive_review'],
 ['On_time_delivery', 'Speed_Slow', 'Positive_review'],
 ['On_time_delivery', 'Speed_Slow', 'Positive_review'],
 ['On_time_delivery', 'Speed_Slow', 'Negative_review'],
 ['On_time_delivery', 'Speed_Slow', 'Positive_review'],
 ['On_time_delivery', 'Speed_Slow', 'Negative_review'],
 ['On_time_delivery', 'Speed_Slow', 'Negative_review'],
 ['On_time_delivery', 'Speed_Normal', 'Positive_review'],
 ['On_time_delivery', 'Speed_Slow', 'Positive_review'],
 ['On_time_delivery', 'Speed_Normal', 'Positive_review'],
 ['On_time_delivery', 'Speed_Slow', 'Positive_review'],
 ['On_time_delivery', 'Speed_Slow', 'Positive_review'],
 ['On_time_delivery', 'Speed_Slow', 'Positive_review'],
 ['On_time_delivery', 'Speed_Normal', 'Posit

In [21]:
import sys
!{sys.executable} -m pip install mlxtend

Defaulting to user installation because normal site-packages is not writeable



[notice] A new release of pip is available: 25.2 -> 25.3
[notice] To update, run: python.exe -m pip install --upgrade pip


In [22]:
from mlxtend.preprocessing import TransactionEncoder

te = TransactionEncoder()
te_array = te.fit(transactions_g1).transform(transactions_g1)

df_trans_g1 = pd.DataFrame(te_array, columns=te.columns_)

In [23]:
df_trans_g1

Late_delivery  Negative_review  On_time_delivery  Positive_review  \
0               False            False              True             True   
1               False            False              True             True   
2               False            False              True             True   
3               False            False              True             True   
4               False            False              True             True   
...               ...              ...               ...              ...   
110000          False            False              True             True   
110001          False            False              True             True   
110002          False             True              True            False   
110003          False             True              True            False   
110004          False            False              True             True   

        Speed_Fast  Speed_Normal  Speed_Slow  Speed_nan  
0            False         False        True      False  
1            False         False        True      False  
2            False         False        True      False  
3            False         False        True      False  
4             True         False       False      False  
...            ...           ...         ...        ...  
110000       False         False        True      False  
110001       False         False        True      False  
110002       False         False        True      False  
110003       False         False        True      False  
110004       False          True       False      False  

[110005 rows x 8 columns]

In [24]:
from mlxtend.frequent_patterns import apriori, association_rules

freq_g1 = apriori(
    df_trans_g1,
    min_support=0.05,
    use_colnames=True
)

In [25]:
rules_g1 = association_rules(
    freq_g1,
    metric='confidence',
    min_threshold=0.6
)

In [26]:
rules_g1_focus = rules_g1[
    rules_g1['consequents'].astype(str).str.contains('review')
].sort_values(by='lift', ascending=False)

In [27]:
rules_g1_focus[['antecedents', 'consequents', 'support', 'confidence', 'lift']].head(10)

antecedents                          consequents  \
16                   (Late_delivery)        (Negative_review, Speed_Slow)   
15       (Late_delivery, Speed_Slow)                    (Negative_review)   
0                    (Late_delivery)                    (Negative_review)   
21                      (Speed_Fast)  (Positive_review, On_time_delivery)   
10                      (Speed_Fast)                    (Positive_review)   
24                    (Speed_Normal)  (Positive_review, On_time_delivery)   
20    (Speed_Fast, On_time_delivery)                    (Positive_review)   
23  (On_time_delivery, Speed_Normal)                    (Positive_review)   
11                    (Speed_Normal)                    (Positive_review)   
5                 (On_time_delivery)                    (Positive_review)   

     support  confidence      lift  
16  0.050316    0.646612  3.709547  
15  0.050316    0.677396  2.920399  
0   0.051161    0.657477  2.834524  
21  0.076487    0.848271  1.144157  
10  0.076987    0.853816  1.111672  
24  0.215536    0.824151  1.111624  
20  0.076487    0.853607  1.111399  
23  0.215536    0.831726  1.082910  
11  0.217436    0.831416  1.082506  
5   0.741394    0.803953  1.046750

Khi đơn hàng giao trễ, khả năng bị đánh giá tiêu cực cao gấp ~2.8 lần so với trung bình.

Giao trễ + giao chậm → nguy cơ đánh giá tiêu cực tăng gần 3 lần.

Đơn hàng trễ hạn thường đi kèm giao chậm và đánh giá xấu
→ cho thấy vấn đề logistics mang tính hệ thống, không phải ngẫu nhiên.

Giao đúng hạn gần như đảm bảo đánh giá tích cực.

Giao nhanh làm tăng nhẹ khả năng đánh giá tích cực, nhưng không mạnh bằng tránh giao trễ.

=> insight:
Trễ hạn	-> Rất mạnh (tiêu cực)

Giao chậm	-> Mạnh

Giao nhanh	-.> Tăng hài lòng nhẹ

Giao đúng hạn -> yêu cầu bắt buộc

=> Kết quả luật kết hợp cho thấy việc cải thiện hiệu quả giao hàng, đặc biệt là giảm tỷ lệ giao trễ, có tác động lớn hơn đến mức độ hài lòng của khách hàng so với việc chỉ tập trung vào tăng tốc độ giao hàng.

## 1.2. Đặc điểm sản phẩm có làm thay đổi mức độ ảnh hưởng của các yếu tố giao hàng đến sự hài lòng của khách hàng hay không?

In [28]:
cols_g2 = [
    'late_item',          # giao hàng trễ / đúng hạn
    'delivery_speed',     # tốc độ giao hàng
    'weight_level',       # đặc điểm sản phẩm
    'price_level',
    'category_item',
    'review_item'         # mức độ hài lòng
]

df_g2 = df[cols_g2].copy()


In [29]:
df_g2 = df_g2.astype(str)

In [30]:
df_g2.isna().sum()

late_item         0
delivery_speed    0
weight_level      0
price_level       0
category_item     0
review_item       0
dtype: int64

In [31]:
transactions_g2 = df_g2.values.tolist()

In [32]:
transactions_g2

[['On_time_delivery',
  'Speed_Slow',
  'Medium',
  'Cheap',
  'Category_housewares',
  'Positive_review'],
 ['On_time_delivery',
  'Speed_Slow',
  'Light',
  'Expensive',
  'Category_perfumery',
  'Positive_review'],
 ['On_time_delivery',
  'Speed_Slow',
  'Medium',
  'Expensive',
  'Category_auto',
  'Positive_review'],
 ['On_time_delivery',
  'Speed_Slow',
  'Medium',
  'Cheap',
  'Category_pet_shop',
  'Positive_review'],
 ['On_time_delivery',
  'Speed_Fast',
  'Light',
  'Cheap',
  'Category_stationery',
  'Positive_review'],
 ['On_time_delivery',
  'Speed_Slow',
  'Heavy',
  'Expensive',
  'Category_auto',
  'Positive_review'],
 ['On_time_delivery',
  'Speed_Slow',
  'Light',
  'Medium',
  'Category_auto',
  'Positive_review'],
 ['On_time_delivery',
  'Speed_Slow',
  'Light',
  'Cheap',
  'Category_furniture_decor',
  'Negative_review'],
 ['On_time_delivery',
  'Speed_Slow',
  'Heavy',
  'Expensive',
  'Category_office_furniture',
  'Positive_review'],
 ['On_time_delivery',
  'Sp

In [33]:
from mlxtend.preprocessing import TransactionEncoder

te = TransactionEncoder()
te_array_g2 = te.fit(transactions_g2).transform(transactions_g2)

df_trans_g2 = pd.DataFrame(te_array_g2, columns=te.columns_)

In [34]:
from mlxtend.frequent_patterns import apriori, association_rules

freq_items_g2 = apriori(
    df_trans_g2,
    min_support=0.03,   # thấp hơn nhóm 1 vì nhiều item hơn
    use_colnames=True
)

In [35]:
rules_g2 = association_rules(
    freq_items_g2,
    metric='confidence',
    min_threshold=0.6
)

In [36]:
rules_g2 = rules_g2[
    rules_g2['consequents'].astype(str).str.contains('review')
]

In [37]:
rules_g2 = rules_g2[
    rules_g2['antecedents'].astype(str).str.contains(
        'Late_delivery|On_time_delivery|Speed_'
    )
]

In [38]:
rules_g2 = rules_g2.sort_values(
    by=['lift', 'confidence'],
    ascending=False
)

rules_g2.head(10)

antecedents                          consequents  \
199                   (Late_delivery)        (Negative_review, Speed_Slow)   
198       (Late_delivery, Speed_Slow)                    (Negative_review)   
39                    (Late_delivery)                    (Negative_review)   
356         (Expensive, Speed_Normal)  (Positive_review, On_time_delivery)   
232                      (Speed_Fast)  (Positive_review, On_time_delivery)   
320               (Cheap, Speed_Fast)  (Positive_review, On_time_delivery)   
390               (Light, Speed_Fast)  (Positive_review, On_time_delivery)   
401              (Medium, Speed_Fast)  (Positive_review, On_time_delivery)   
425  (Expensive, Heavy, Speed_Normal)  (Positive_review, On_time_delivery)   
374             (Heavy, Speed_Normal)  (Positive_review, On_time_delivery)   

     antecedent support  consequent support   support  confidence      lift  \
199            0.077815            0.174310  0.050316    0.646612  3.709547   
198            0.074278            0.231953  0.050316    0.677396  2.920399   
39             0.077815            0.231953  0.051161    0.657477  2.834524   
356            0.076797            0.741394  0.065188    0.848840  1.144925   
232            0.090169            0.741394  0.076487    0.848271  1.144157   
320            0.040344            0.741394  0.034153    0.846553  1.141840   
390            0.042262            0.741394  0.035680    0.844268  1.138758   
401            0.043007            0.741394  0.036253    0.842951  1.136981   
425            0.040635            0.741394  0.034108    0.839374  1.132157   
374            0.078424            0.741394  0.064797    0.826243  1.114446   

     representativity  leverage  conviction  zhangs_metric   jaccard  \
199               1.0  0.036752    2.336497       0.792059  0.249324   
198               1.0  0.033087    2.380771       0.710344  0.196611   
39                1.0  0.033112    2.242320       0.701819  0.197835   
356               1.0  0.008252    1.710812       0.137110  0.086571   
232               1.0  0.009637    1.704397       0.138481  0.101298   
320               1.0  0.004242    1.685309       0.129442  0.045684   
390               1.0  0.004348    1.660582       0.127227  0.047702   
401               1.0  0.004368    1.646658       0.125892  0.048457   
425               1.0  0.003981    1.609987       0.121674  0.045603   
374               1.0  0.006654    1.488324       0.111432  0.085822   

     certainty  kulczynski  
199   0.572009    0.467635  
198   0.579968    0.447159  
39    0.554033    0.439022  
356   0.415482    0.468383  
232   0.413282    0.475719  
320   0.406637    0.446309  
390   0.397801    0.446197  
401   0.392709    0.445925  
425   0.378877    0.442689  
374   0.328103    0.456821

Mặc dù luật tiêu cực chủ yếu xoay quanh giao hàng, việc các luật tích cực phía dưới đều gắn với đặc điểm sản phẩm cho thấy:

Khi điều kiện giao hàng không thuận lợi, đặc điểm sản phẩm không đủ để “cứu” trải nghiệm, và đánh giá tiêu cực vẫn chiếm ưu thế.

Điều này gián tiếp khẳng định vai trò điều chỉnh (chứ không thay thế) của sản phẩm.

Khi giao hàng thuận lợi, các đặc điểm sản phẩm:

nhẹ / nặng

rẻ / trung bình / đắt

đều duy trì được mức độ hài lòng cao.

Tác động của giao hàng mạnh hơn sản phẩm rất nhiều
Sản phẩm không tạo ra đột biến, mà:

giúp củng cố trải nghiệm tích cực khi giao hàng đã tốt

đặc điểm sản phẩm không phải là yếu tố quyết định chính gây ra sự không hài lòng của khách hàng. Tuy nhiên, khi kết hợp với điều kiện giao hàng thuận lợi, các đặc điểm sản phẩm như trọng lượng và mức giá giúp duy trì mức độ hài lòng cao. Ngược lại, trong các trường hợp giao hàng trễ hoặc giao chậm, tác động tiêu cực của giao hàng chi phối mạnh mẽ trải nghiệm khách hàng, bất kể đặc điểm sản phẩm

Mức độ hài lòng của khách hàng trên nền tảng Olist chịu ảnh hưởng chủ yếu từ yếu tố giao hàng. Đặc điểm sản phẩm đóng vai trò điều chỉnh, làm rõ sự khác biệt trong trải nghiệm khi điều kiện giao hàng thuận lợi, nhưng không đủ để bù đắp cho các vấn đề giao hàng.

## 1.3. Khách hàng có sẵn sàng chấp nhận chi phí cao nếu trải nghiệm giao hàng tốt không?

In [39]:
transactions_g3 = df[[
    'price_level',        # Cheap / Medium / Expensive
    'freight_level',      # Low / Medium / High
    'delivery_speed',     # Fast / Normal / Slow
    'late_item',          # On_time_delivery / Late_delivery
    'review_item'         # Positive_review / Negative_review
]].astype(str).values.tolist()

In [40]:
transactions_g3

[['Cheap', 'Low', 'Speed_Slow', 'On_time_delivery', 'Positive_review'],
 ['Expensive', 'High', 'Speed_Slow', 'On_time_delivery', 'Positive_review'],
 ['Expensive', 'High', 'Speed_Slow', 'On_time_delivery', 'Positive_review'],
 ['Cheap', 'High', 'Speed_Slow', 'On_time_delivery', 'Positive_review'],
 ['Cheap', 'Low', 'Speed_Fast', 'On_time_delivery', 'Positive_review'],
 ['Expensive', 'High', 'Speed_Slow', 'On_time_delivery', 'Positive_review'],
 ['Medium', 'Medium', 'Speed_Slow', 'On_time_delivery', 'Positive_review'],
 ['Cheap', 'Medium', 'Speed_Slow', 'On_time_delivery', 'Negative_review'],
 ['Expensive', 'High', 'Speed_Slow', 'On_time_delivery', 'Positive_review'],
 ['Medium', 'High', 'Speed_Slow', 'On_time_delivery', 'Negative_review'],
 ['Medium', 'High', 'Speed_Slow', 'On_time_delivery', 'Negative_review'],
 ['Medium', 'Medium', 'Speed_Normal', 'On_time_delivery', 'Positive_review'],
 ['Cheap', 'Medium', 'Speed_Slow', 'On_time_delivery', 'Positive_review'],
 ['Cheap', 'Low', 'Spee

In [41]:
from mlxtend.preprocessing import TransactionEncoder

te = TransactionEncoder()
te_array = te.fit(transactions_g3).transform(transactions_g3)

df_trans_g3 = pd.DataFrame(te_array, columns=te.columns_)

In [42]:
from mlxtend.frequent_patterns import apriori, association_rules

frequent_itemsets_g3 = apriori(
    df_trans_g3,
    min_support=0.03,
    use_colnames=True
)

In [43]:
rules_g3 = association_rules(
    frequent_itemsets_g3,
    metric='confidence',
    min_threshold=0.7
)

In [44]:
rules_g3_focus = rules_g3[
    rules_g3['antecedents'].astype(str).str.contains(
        'Cheap|Medium|Expensive|Low|High', regex=True
    ) &
    rules_g3['consequents'].astype(str).str.contains(
        'Positive_review|Negative_review|On_time_delivery|Late_delivery',
        regex=True
    )
]


In [45]:
rules_g3_focus = rules_g3_focus.sort_values(
    by=['lift', 'confidence'],
    ascending=False
)

rules_g3_focus[[
    'antecedents',
    'consequents',
    'support',
    'confidence',
    'lift'
]].head(15)

antecedents  \
219                    (Cheap, Speed_Fast)   
218  (Cheap, Speed_Fast, On_time_delivery)   
128                    (Cheap, Speed_Fast)   
216   (Positive_review, Cheap, Speed_Fast)   
123                    (Cheap, Speed_Fast)   
188                (Positive_review, High)   
240        (Positive_review, High, Medium)   
232     (Positive_review, Expensive, High)   
117                          (High, Cheap)   
228       (Positive_review, Cheap, Medium)   
72                                  (High)   
177                         (High, Medium)   
200                      (Low, Speed_Fast)   
217               (Low, Cheap, Speed_Fast)   
168              (Expensive, Speed_Normal)   

                                  consequents   support  confidence      lift  
219  (Positive_review, Low, On_time_delivery)  0.031644    0.784362  3.083656  
218                    (Positive_review, Low)  0.031644    0.788270  2.993118  
128                    (Positive_review, Low)  0.031799    0.788193  2.992826  
216                   (Low, On_time_delivery)  0.031644    0.922119  2.936904  
123                   (Low, On_time_delivery)  0.036898    0.914601  2.912959  
188            (Speed_Slow, On_time_delivery)  0.184255    0.738048  1.287426  
240            (Speed_Slow, On_time_delivery)  0.053579    0.737857  1.287093  
232            (Speed_Slow, On_time_delivery)  0.103904    0.730118  1.273594  
117            (Speed_Slow, On_time_delivery)  0.034735    0.726012  1.266431  
228            (Speed_Slow, On_time_delivery)  0.067951    0.709203  1.237110  
72             (Speed_Slow, On_time_delivery)  0.234362    0.703726  1.227556  
177            (Speed_Slow, On_time_delivery)  0.068824    0.700889  1.222607  
200       (Positive_review, On_time_delivery)  0.061752    0.853714  1.151499  
217       (Positive_review, On_time_delivery)  0.031644    0.853186  1.150787  
168       (Positive_review, On_time_delivery)  0.065188    0.848840  1.144925

Chi phí thấp không tự tạo ra hài lòng, nhưng khi kết hợp với giao hàng nhanh, nó tạo thành một tổ hợp trải nghiệm rất tích cực.

Khách hàng chấp nhận giao chậm nếu không trễ, nhưng phí cao + giao chậm làm trải nghiệm kém hấp dẫn hơn so với tổ hợp chi phí thấp – giao nhanh.

Giao hàng vẫn là yếu tố nền

Chi phí chỉ điều chỉnh mức độ hài lòng, không quyết định tuyệt đối

Mức độ hài lòng của khách hàng trên Olist không phụ thuộc riêng lẻ vào chi phí hay giao hàng, mà hình thành từ các tổ hợp yếu tố.

## 1.4. KHU VỰC KHÁCH HÀNG TRONG MỐI LIÊN HỆ VỚI GIAO HÀNG & HÀI LÒNG

In [46]:
transactions_g4 = df[
    [
        'state_item',
        'late_item',
        'delivery_speed',
        'review_item'
    ]
].values.tolist()

In [47]:
transactions_g4[:5]

[['State_SP', 'On_time_delivery', 'Speed_Slow', 'Positive_review'],
 ['State_BA', 'On_time_delivery', 'Speed_Slow', 'Positive_review'],
 ['State_GO', 'On_time_delivery', 'Speed_Slow', 'Positive_review'],
 ['State_RN', 'On_time_delivery', 'Speed_Slow', 'Positive_review'],
 ['State_SP', 'On_time_delivery', 'Speed_Fast', 'Positive_review']]

In [48]:
from mlxtend.preprocessing import TransactionEncoder

te = TransactionEncoder()
te_array = te.fit(transactions_g4).transform(transactions_g4)

df_trans_g4 = pd.DataFrame(te_array, columns=te.columns_)

In [49]:
from mlxtend.frequent_patterns import apriori

frequent_itemsets_g4 = apriori(
    df_trans_g4,
    min_support=0.01,
    use_colnames=True
)

In [50]:
from mlxtend.frequent_patterns import association_rules

rules_g4 = association_rules(
    frequent_itemsets_g4,
    metric='confidence',
    min_threshold=0.6
)

In [51]:
rules_g4_filtered = rules_g4[
    rules_g4['antecedents'].apply(lambda x: any('State_' in i for i in x)) &
    rules_g4['consequents'].apply(lambda x: any('review' in i.lower() for i in x))
]

In [52]:
rules_g4_filtered = rules_g4_filtered.sort_values(
    by=['lift', 'confidence'],
    ascending=False
)

rules_g4_filtered.head(10)

antecedents  \
191              (State_RJ, Late_delivery)   
190  (State_RJ, Late_delivery, Speed_Slow)   
53               (State_RJ, Late_delivery)   
193  (Late_delivery, State_SP, Speed_Slow)   
263           (State_PE, On_time_delivery)   
232           (State_BA, On_time_delivery)   
264                             (State_PE)   
251           (State_GO, On_time_delivery)   
178                             (State_PE)   
244           (State_ES, On_time_delivery)   

                                         consequents  antecedent support  \
191                    (Negative_review, Speed_Slow)            0.016363   
190                                (Negative_review)            0.016081   
53                                 (Negative_review)            0.016363   
193                                (Negative_review)            0.021254   
263                    (Positive_review, Speed_Slow)            0.014236   
232                    (Positive_review, Speed_Slow)            0.028871   
264  (Positive_review, Speed_Slow, On_time_delivery)            0.015817   
251                    (Positive_review, Speed_Slow)            0.018972   
178                    (Positive_review, Speed_Slow)            0.015817   
244                    (Positive_review, Speed_Slow)            0.017654   

     consequent support   support  confidence      lift  representativity  \
191            0.174310  0.012527    0.765556  4.391913               1.0   
190            0.231953  0.012527    0.778971  3.358313               1.0   
53             0.231953  0.012609    0.770556  3.322032               1.0   
193            0.231953  0.012981    0.610778  2.633198               1.0   
263            0.473242  0.010572    0.742656  1.569295               1.0   
232            0.473242  0.020572    0.712531  1.505638               1.0   
264            0.449280  0.010572    0.668391  1.487695               1.0   
251            0.473242  0.013281    0.700048  1.479260               1.0   
178            0.473242  0.011018    0.696552  1.471872               1.0   
244            0.473242  0.012145    0.687951  1.453697               1.0   

     leverage  conviction  zhangs_metric   jaccard  certainty  kulczynski  
191  0.009674    3.521899       0.785156  0.070317   0.716062    0.418710  
190  0.008797    3.474872       0.713709  0.053190   0.712220    0.416488  
53   0.008813    3.347420       0.710607  0.053492   0.701262    0.412457  
193  0.008051    1.973290       0.633702  0.054038   0.493232    0.333372  
263  0.003835    2.046905       0.368010  0.022168   0.511458    0.382498  
232  0.006909    1.832402       0.345814  0.042721   0.454268    0.378001  
264  0.003466    1.660751       0.333088  0.023260   0.397863    0.345961  
251  0.004303    1.756140       0.330252  0.027731   0.430569    0.364056  
178  0.003532    1.735907       0.325745  0.023048   0.423932    0.359917  
244  0.003790    1.688059       0.317707  0.025368   0.407604    0.356807

Ở các bang lớn như RJ, SP, khi giao chậm + trễ, xác suất đánh giá tiêu cực tăng gấp 3–4 lần

Lift rất cao ⇒ mối liên hệ mạnh, không phải ngẫu nhiên

Cùng là giao trễ, nhưng ở một số khu vực, tác động tiêu cực mạnh hơn rõ rệt.

Giao chậm nhưng đúng hạn vẫn thường đi kèm Positive_review

Điều này xuất hiện ổn định ở nhiều bang

Khu vực khách hàng không phải nguyên nhân trực tiếp, mà là yếu tố điều kiện làm thay đổi mức độ tác động của giao hàng đến sự hài lòng.

Khu vực khách hàng không ảnh hưởng trực tiếp đến mức độ hài lòng, nhưng đóng vai trò như một yếu tố bối cảnh quan trọng. Khi kết hợp với trạng thái và tốc độ giao hàng, khu vực địa lý có thể làm gia tăng hoặc làm giảm tác động của giao hàng đến đánh giá của khách hàng.

## 1.5. Tổng kết lại

In [53]:
transactions_g5 = df[
    [
        # Delivery
        'late_item',
        'delivery_speed',

        # Product
        'category_item',
        'weight_level',

        # Cost
        'price_level',
        'freight_level',

        # Region
        'state_item',

        # Output
        'review_item'
    ]
].values.tolist()

In [54]:
transactions_g5[:5]

[['On_time_delivery',
  'Speed_Slow',
  'Category_housewares',
  'Medium',
  'Cheap',
  'Low',
  'State_SP',
  'Positive_review'],
 ['On_time_delivery',
  'Speed_Slow',
  'Category_perfumery',
  'Light',
  'Expensive',
  'High',
  'State_BA',
  'Positive_review'],
 ['On_time_delivery',
  'Speed_Slow',
  'Category_auto',
  'Medium',
  'Expensive',
  'High',
  'State_GO',
  'Positive_review'],
 ['On_time_delivery',
  'Speed_Slow',
  'Category_pet_shop',
  'Medium',
  'Cheap',
  'High',
  'State_RN',
  'Positive_review'],
 ['On_time_delivery',
  'Speed_Fast',
  'Category_stationery',
  'Light',
  'Cheap',
  'Low',
  'State_SP',
  'Positive_review']]

In [55]:
from mlxtend.preprocessing import TransactionEncoder

te = TransactionEncoder()
te_array = te.fit(transactions_g5).transform(transactions_g5)

df_trans_g5 = pd.DataFrame(te_array, columns=te.columns_)

In [56]:
from mlxtend.frequent_patterns import apriori

frequent_itemsets_g5 = apriori(
    df_trans_g5,
    min_support=0.03,
    use_colnames=True
)

In [57]:
from mlxtend.frequent_patterns import association_rules

rules_g5 = association_rules(
    frequent_itemsets_g5,
    metric='confidence',
    min_threshold=0.6
)

In [58]:
def count_groups(items):
    groups = {
        'delivery': ['Late_delivery', 'On_time_delivery', 'Speed_'],
        'product': ['Category_', 'Light', 'Medium', 'Heavy'],
        'cost': ['Cheap', 'Medium', 'Expensive', 'Low', 'High'],
        'region': ['State_']
    }

    used = set()
    for item in items:
        for g, keys in groups.items():
            if any(k in item for k in keys):
                used.add(g)
    return len(used)

rules_g5_filtered = rules_g5[
    rules_g5['consequents'].apply(lambda x: any('review' in i.lower() for i in x))
]

rules_g5_filtered['num_groups'] = rules_g5_filtered['antecedents'].apply(count_groups)

rules_g5_final = rules_g5_filtered[
    (rules_g5_filtered['num_groups'] >= 3) &
    (rules_g5_filtered['lift'] > 1.2) &
    (rules_g5_filtered['support'] > 0.01)
]

C:\Users\biena\AppData\Local\Temp\ipykernel_16544\2894498882.py:20: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  rules_g5_filtered['num_groups'] = rules_g5_filtered['antecedents'].apply(count_groups)


In [59]:
rules_g5_final = rules_g5_final.sort_values(
    by=['lift', 'confidence'],
    ascending=False
)

rules_g5_final.head(10)

antecedents  \
1001                            (Late_delivery, Medium)   
999                 (Late_delivery, Speed_Slow, Medium)   
360                             (Late_delivery, Medium)   
1769         (Light, State_SP, Cheap, On_time_delivery)   
1831                    (Light, State_SP, Speed_Normal)   
1808                    (State_SP, Cheap, Speed_Normal)   
1774                           (Light, State_SP, Cheap)   
1623                    (Light, State_SP, Speed_Normal)   
1828  (Light, On_time_delivery, State_SP, Speed_Normal)   
1322                           (Light, State_SP, Cheap)   

                                   consequents  antecedent support  \
1001             (Negative_review, Speed_Slow)            0.051379   
999                          (Negative_review)            0.049461   
360                          (Negative_review)            0.051379   
1769                    (Positive_review, Low)            0.094278   
1831  (Positive_review, Low, On_time_delivery)            0.064006   
1808  (Positive_review, Low, On_time_delivery)            0.064806   
1774  (Positive_review, Low, On_time_delivery)            0.099259   
1623                    (Positive_review, Low)            0.064006   
1828                    (Positive_review, Low)            0.062988   
1322                    (Positive_review, Low)            0.099259   

      consequent support   support  confidence      lift  representativity  \
1001            0.174310  0.033644    0.654812  3.756592               1.0   
999             0.231953  0.033644    0.680206  2.932515               1.0   
360             0.231953  0.034080    0.663305  2.859652               1.0   
1769            0.263361  0.066915    0.709768  2.695039               1.0   
1831            0.254361  0.043807    0.684420  2.690740               1.0   
1808            0.254361  0.043771    0.675410  2.655320               1.0   
1774            0.254361  0.066915    0.674146  2.650349               1.0   
1623            0.263361  0.044562    0.696208  2.643552               1.0   
1828            0.263361  0.043807    0.695483  2.640799               1.0   
1322            0.263361  0.068997    0.695119  2.639416               1.0   

      leverage  conviction  zhangs_metric   jaccard  certainty  kulczynski  \
1001  0.024688    2.392003       0.773546  0.175187   0.581940    0.423912   
999   0.022171    2.401692       0.693287  0.135787   0.583627    0.412626   
360   0.022163    2.281136       0.685529  0.136730   0.561622    0.405116   
1769  0.042086    2.538102       0.694416  0.230168   0.606005    0.481925   
1831  0.027526    2.362756       0.671324  0.159554   0.576765    0.428322   
1808  0.027287    2.297173       0.666597  0.158937   0.564682    0.423746   
1774  0.041667    2.288260       0.691310  0.233394   0.562987    0.468609   
1623  0.027705    2.424814       0.664236  0.157570   0.587597    0.432706   
1828  0.027219    2.419039       0.663094  0.155046   0.586613    0.430911   
1322  0.042856    2.416150       0.689575  0.234985   0.586118    0.478552   

      num_groups  
1001           3  
999            3  
360            3  
1769           4  
1831           3  
1808           3  
1774           3  
1623           3  
1828           3  
1322           3

Late_delivery là lõi tiêu cực

Khi kết hợp thêm:

Sản phẩm nặng

Chi phí cao / trung bình
→ Xác suất đánh giá tiêu cực tăng gấp ~4 lần

Kết luận quan trọng:

Khi giao hàng trễ xảy ra, các yếu tố sản phẩm và chi phí khuếch đại mạnh tác động tiêu cực lên sự hài lòng.

Giao nhanh + đúng hạn là nền

Giá rẻ + phí thấp + hàng nhẹ

Khu vực SP (gần trung tâm logistics)

→ Tạo thành tổ hợp trải nghiệm gần như lý tưởng

Ý nghĩa:

Không yếu tố nào trong số này đủ mạnh một mình, nhưng khi kết hợp lại → mức độ hài lòng tăng vọt (lift > 3).

State_SP chỉ xuất hiện trong các luật tích cực

Không xuất hiện đơn lẻ, mà luôn đi kèm:

Speed_Fast

On_time_delivery

Low freight

Khu vực không tạo hài lòng, nhưng tạo điều kiện để giao hàng hiệu quả, từ đó hình thành trải nghiệm tích cực.

Sự hài lòng không đến từ một yếu tố đơn lẻ

Các tổ hợp tích cực (giao nhanh, đúng hạn, chi phí thấp, hàng nhẹ, khu vực thuận lợi) làm xác suất đánh giá tích cực tăng gấp nhiều lần

Kết quả khai phá luật kết hợp cho thấy mức độ hài lòng của khách hàng trên nền tảng Olist được hình thành từ sự tương tác của nhiều yếu tố liên quan đến giao hàng, sản phẩm, chi phí và khu vực địa lý. Không có yếu tố nào tác động độc lập, mà chính các tổ hợp điều kiện mới quyết định trải nghiệm và đánh giá của khách hàng.